Fix for RGB/RGBA Colors using new notation and documentation (fixes issue #476)

WilliamJohnsonJr opened the pull request:
### Description of change

Fixes issue [#476 RGBA notation doesn't work for colouring pie chart](https://github.com/jtblin/angular-chart.js/issues/476)

Adds a new notation for RGB/RGBA colors, and new documentation for how to input RGB/RGBA colors for a chart.

Added getRGBAColor function to ChartJSFactory, added new if statements to convertColor in ChartJSFactory so that RGB colors, RGBA colors, and colors input as objects would be properly applied to charts. Wrote an appropriate test to check for each type of color input format in test.unit.js. Updated documentation regarding colors in README.md. Input for colors is different from standard RGBA/RGB Colors notation.

Note: When running 'gulp check' all tests pass, but no integrations pass. Unsure why. The fix works on projects I'm running on my local machine, and apparently for test.unit.js. Also: files in the 'dist' folder remain unchanged. I have only changed 'angular-chart.js' in the main project folder.
### Pull Request check-list
- [x] Run `gulp test` to ensure there are no linting, or style issues and all tests pass.
- [x] Squash your commits into a few commits only.
- [x] Make sure the commit message is short, concise and descriptive of the issues you're fixing.
- [x] Avoid mixing up multiple issues and/or features, open one pull request for each issue.
- [x] Have you updated the documentation and / or examples?
- [x] Have you included a new test?

jtblin commented on angular-chart.js:
```diff
@@ -169,7 +169,7 @@
       scope.chartGetColor = getChartColorFn(scope);
       var data = getChartData(type, scope);
-
+      console.log(data);
```
Please remove the `console.log` statement.

jtblin commented on angular-chart.js:
```diff
@@ -277,6 +295,7 @@
 
     function getChartData (type, scope) {
       var colors = getColors(type, scope);
+      console.log(colors);
```
Please remove the `console.log` statement.

jtblin commented on angular-chart.js:
```diff
@@ -227,8 +227,15 @@
+      // Allows hex colors to be input as a string.
       if (typeof color === 'string' && color[0] === '#') return getColor(hexToRgb(color.substr(1)));
+      // Allows colors to be input as an object, bypassing getColor() entirely
+      if (typeof color === 'object' && color !==null) return color;
```
Nit: space after `!==`

jtblin commented:
Thanks a lot for the PR @WilliamJohnsonJr! 

IIUC #476, the user wanted to pass an `rgba` string directly in the colors, I don't think your PR would allow this e.g.

```
$scope.arrayColors = [
    "rgba(77,83,96,0.5)",
    "rgba(253,180,92,0.5)",
    "rgba(175, 101, 255,0.5)",
    "rgba(255, 40, 44, 0.5)",
    "rgba(96, 255, 173, 0.5)",
    "rgba(220,220,220,0.5)"
  ];
```
Maybe handle different kind of strings?

Another thing is that `getRGBAColor` seems to be doing almost the same thing as `getColor`. I wonder if it would be possible to reuse `getColor` instead of duplicating it but maybe I'm missing something. For example test if the length of color is 3 or 4 in `getColor` instead and use this to set the tones accordingly. Also you are doing `color.slice(0,3)` multiple times, it would be better to store the value in a variable and use that instead. Something like this?

```
function getColor (color) {
  var alpha = color[3];
  color = color.slice(0, 3);
  return {
    backgroundColor: rgba(color, 0.2),
    pointBackgroundColor: rgba(color, alpha || 1),
    pointHoverBackgroundColor: rgba(color, 0.8),
    borderColor: rgba(color, alpha || 1),
    pointBorderColor: '#fff',
    pointHoverBorderColor: rgba(color, alpha || 1)
  };
}
```

WilliamJohnsonJr commented:
Thanks for the feedback @jtblin ! I removed getRGBAColor, added the requested functionality into getColor(), and updated the README and tests. Let me know what you think!

jtblin commented on angular-chart.js:
```diff
@@ -238,15 +241,53 @@
+      if (typeof color === 'object'){
+        return colorSetter(color, 1);
+      }
+      if (typeof color === 'string'){
```
All this logic should be better placed in a different function, to avoid creating the inner function `colorSetter` above e.g. `rgbStringToRgb`.

jtblin commented on angular-chart.js:
```diff
@@ -238,15 +241,53 @@
+        // the RGBA or RGB string into an array for processing by the rgba() function when it tests the browser for RGBA color support.
+        var rgbaArray=[];
+        var commaIndexes = [];
+        var position = color.indexOf(',');
```
All the logic below can be simplified to a simple `color.split(',')` which will return an array.

jtblin commented on angular-chart.js:
```diff
@@ -238,15 +241,53 @@
+          rgbaArray.push(color.substring(commaIndexes[2]+1, color.length-1));  
+        };         
+        color = rgbaArray.slice(0,3);
+        var aVal = rgbaArray[3];
```
`aVal` is too generic, it's the alpha element so it would be better named `alpha`.

jtblin replied:
Instead of testing the length etc. you can just used the fact that `rgbaArray[3]` will return `undefined` if it doesn't exist and set it to 1 instead so basically the entire logic can be changed to:

```
function getColor (color) {
   color = color.slice(0, 3);
   var alpha = color[3] || 1;
   return { ... };
}

function rgbStringToRgb (color) {
    return color.split(',');
}
```

WilliamJohnsonJr commented:
Thanks again for the feedback @jtblin ! I've updated the code accordingly. Glad to make further updates if you have any feedback on these commits.

jtblin commented on angular-chart.js:
```diff
@@ -267,6 +272,23 @@
       return [r, g, b];
     }
 
+    function rgbStringToRgb (color) {
```
This can be greatly simplified by using regexp and the `Number` javascript core function which parses a string into a Number, e.g.

```
function rgbStringToRgb (color) { 
  var match = color.match(/^rgba?\(([\d,\.]+)\)$/);
  if (! match) throw new Error('Cannot parse rgb value');
  color = match[1].split(','); 
  return color.map(Number);
}
```

WilliamJohnsonJr commented:
@jtblin : Thanks for the feedback! I have updated rgbStringToRgb using RegExp and the Number method as requested.

uofmmike commented:
Thanks for the work on this @WilliamJohnsonJr, just ran into this now! Patiently awaiting the merge approval!

oqx commented:
Would love to see this merged. 👯

WilliamJohnsonJr commented:
@jtblin : Hope the recent update is suitable to pull in. Is there anything else I should update for this pull request to go through? Glad to make it happen if there's something specific to fix.

jtblin commented:
Sorry for the lag, LGTM. I will have a last look and merge this weekend. Thank you for your patience!
